src/application-manager.coffee

```diff
@@ -0,0 +1,875 @@
+					steps = steps.concat(pairSteps) if !_.isEmpty(pairSteps)
+				_.forEach imagePairs, (pair) ->
+					steps.push(_.assign({ action: 'updateImage' }, pair))
+				return steps
+		)
+
+	normaliseAppForDB: (app) =>
+		services = _.map app.services, (s, serviceId) ->
+			service = _.clone(s)
+			service.appId = app.appId
+			service.releaseId = app.releaseId
+			service.serviceId = serviceId
+			service.commit = app.commit
+			return service
+		Promise.map services, (service) =>
+			service.image = @images.normalise(service.image)
+			Promise.props(service)
+		.then (services) ->
+			dbApp = {
+				appId: app.appId
+				commit: app.commit
+				name: app.name
+				releaseId: app.releaseId
+				services: JSON.stringify(services)
+				networks: JSON.stringify(app.networks ? {})
+				volumes: JSON.stringify(app.volumes ? {})
+			}
+			return dbApp
+
+	createTargetService: (service, opts) ->
+		NotFoundErr = (err) -> err.statusCode == 404
+		serviceOpts = {
+			serviceName: service.serviceName
+		}
+		_.assign(serviceOpts, opts)
+		Promise.join(
+			@images.inspectByName(service.image)
+			.catchReturn(undefined)
+			@docker.getNetworkGateway(service.network_mode ? service.appId)
+			.catchReturn(NotFoundErr, null)
+			.catchReturn(@docker.InvalidNetGatewayError, null)
+			(imageInfo, apiHostForNetwork) ->
+				serviceOpts.imageInfo = imageInfo
+				serviceOpts.supervisorApiHost = apiHostForNetwork if apiHostForNetwork?
+				return new Service(service, serviceOpts)
+		)
+
+	normaliseAndExtendAppFromDB: (app) =>
+		Promise.join(
+			@config.get('extendedEnvOptions')
+			@docker.defaultBridgeGateway()
+			(opts, supervisorApiHost) =>
+				configOpts = {
+					appName: app.name
+					commit: app.commit
+					supervisorApiHost
+				}
+				_.assign(configOpts, opts)
+				volumes = JSON.parse(app.volumes)
+				volumes = _.mapValues volumes, (volumeConfig) ->
+					volumeConfig ?= {}
+					volumeConfig.labels ?= {}
+					return volumeConfig
+				Promise.map(JSON.parse(app.services), (service) => @createTargetService(service, configOpts))
+				.then (services) ->
+					# If a named volume is defined in a service, we add it app-wide so that we can track it and purge it
+					_.forEach services, (s) ->
+						serviceNamedVolumes = s.getNamedVolumes()
+						_.forEach serviceNamedVolumes, (name) ->
+							volumes[name] ?= { labels: {} }
+					outApp = {
+						appId: app.appId
+						name: app.name
+						commit: app.commit
+						releaseId: app.releaseId
+						services: services
+						networks: JSON.parse(app.networks)
+						volumes: volumes
+					}
+					return outApp
+		)
+
+	setTarget: (apps, dependent , trx) =>
+		setInTransaction = (trx) =>
+			Promise.try =>
+				if apps?
+					appsArray = _.map apps, (app, appId) ->
+						appClone = _.clone(app)
+						appClone.appId = appId
+						return appClone
+					Promise.map(appsArray, @normaliseAppForDB)
+					.then (appsForDB) =>
```

Comment:
You can do `.map` directly here

Reply:
Not really, cause we also use appsForDB as a collection in line 687 below (we could `return app` inside the map, but then it kinda defeats the purpose of using the .map)